proxmox: Add public key support: fix #19487

DavidVentura opened the pull request:
##### ISSUE TYPE
Bugfix Pull Request

##### COMPONENT NAME
proxmox

##### ANSIBLE VERSION
2.3

##### SUMMARY

this PR fixes #19487 (adding pubkey support to proxmox module)

Etherdaemon committed:
New facts module for ec2 nat gateway facts module

Etherdaemon committed:
Update typo for connection to the get_nat_gateways function

Etherdaemon committed:
Remove camel_dict_to_snake_dict as this was breaking the module itself - the result is a list

bcoca committed:
added warnings list to module and autoadd

added better way of adding warnings to return data
backwards compatible if warnings key already exists
added deprecations made iface more generic
changed to enforce type per item
added logging of warnings/deprecations
also display deprecations by default

stevekuznetsov committed:
ec2_group_facts: Fail correctly when boto3 is not installed

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

bcoca committed:
refactoring async

- centralized skipping
- also fixed module name broken by previous refactor
- let action modules handle async processing
- moved async into base action class's module exec
- action plugins can now run final action as async
- actually skip copy if base skips
- fixed normal for new paths
- ensure internal stat is never async
- default poll to 10 as per docs
- added hint for callback fix on poll
- restructured late tmp, now a pipeline query
- moving action handler to connection as networking does
- fixed network assumption invocation is always passed
- centralized key cleanup, normalized internal var
- _supress_tmpdir_delete now in _ansible_xxx and gets removed from results
- delay internal key removal till after we use em
- nicer tmp removing, using existing methods
- moved cleanup tmp flag to mking tmp func

DavidVentura committed:
fix #19487

DavidVentura committed:
add 'version_added'

gundalow commented on lib/ansible/modules/cloud/misc/proxmox.py:
```diff
@@ -175,6 +175,14 @@
+  pubkey:
+    version_added: "2.3"
+    description:
+      - Public key to add to /root/.ssh/authorized_keys
```
`Public key to add to .ssh/authorized_keys.`
Missing full stop
May not be run as root

DavidVentura replied:
Why can't it be run as root?

gundalow replied:
Apologies, my comment wasn't clear.
I assume this module could be ran as root *or* another user. Therefore the `description:` should be something closer to
`Public key to add to users .ssh/authorized_keys.

DavidVentura replied:
Proxmox adds the specified public key to the root user upon container creation.

gundalow commented on lib/ansible/modules/cloud/misc/proxmox.py:
```diff
@@ -175,6 +175,14 @@
     default: present
+  pubkey:
+    version_added: "2.3"
+    description:
```
Generally we put the `description:` first to make the file more readable.

DavidVentura replied:
Swapped in the last commit

gundalow reviewed (changes requested):
needs_revision

UnderGreen commented:
Will fail on ProxmoxVE < 4.2. You have to add backport compatibility.

DavidVentura commented:
@UnderGreen It will only fail **only** if the options are specified and proxmox does not support it. You get an api response saying that the option is not supported. Is this behavior not ok?

DavidVentura commented:
Proxmox adds the key to the root user upon creation

On December 19, 2016 1:52:17 PM GMT-03:00, John R Barker <notifications@github.com> wrote:
>gundalow commented on this pull request.
>
>
>
>> @@ -175,6 +175,14 @@
>      - Indicate desired state of the instance
>     choices: ['present', 'started', 'absent', 'stopped', 'restarted']
>     default: present
>+  pubkey:
>+    version_added: "2.3"
>+    description:
>+      - Public key to add to /root/.ssh/authorized_keys
>
>Apologies, my comment wasn't clear.
>I assume this module could be ran as root *or* another user. Therefore
>the `description:` should be something closer to
>`Public key to add to users .ssh/authorized_keys.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/ansible/ansible/pull/19515

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

UnderGreen commented:
@DavidVentura if Proxmox doesn't support something it will response with 500x :cry:

DavidVentura committed:
Check for version compatibility, Ignore keys if incompatible

DavidVentura commented:
@UnderGreen Just pushed the version check

alikins committed:
Show valid bool symbols in error for invalid bool symbol

simplesteph committed:
added deployment configuration as an option. this fixes #2434

simplesteph committed:
moved utils function to utils modules, fixed a small typo

Commit:
Fix python3, fix flake8, use common code

Update exception handling, remove use of iteritems
Update for better flake8 compliance

Use ansible_dict_to_boto3_filter_list rather than
duplicating its implementation

n0trax committed:
Add workaround for shib2 and php modules

Commit:
Add support for ClearLinux

* Set ansible_os_family to 'ClearLinux'
* Set ansible_distribution to 'ClearLinux'
* Add swupd.py plugin

Fixes #20824

Signed-off-by: Alberto Murillo Silva <alberto.murillo.silva@intel.com>

tedder committed:
add message_attributes support to sns.

bcoca committed:
allow empty whitelist

aperigault committed:
vmware_guest: Fix a bug when deploy a VM is not always considered as a change

dagwieers committed:
Small cosmetic changes and documentation updates (#20632)

After merging the latest update, these were the few remaining
differences with upstream.

(Plus the previous proposed fast match based on unique VM name which is
not longer needed)

dagwieers committed:
Fix the integration test to comply recent changes

mattclay committed:
Temporarily switch Windows tests to us-east-1.

Will switch back to us-east-2 after AWS limit increase is done.

evgeni committed:
redhat_subscription: Document & enforce org_id (#20548)

* redhat_subscription: enforce the need for org_id when AK is used

also update the documentation to reflect that

Fixes: #20542

* redhat_subscription: check for existence of subscription-manager

Commit:
[redhat_subscription]: fixed force option. (#20667)

`force` option is not unique for non-activationkey registrations and it should
be parsed separately; it is possible to register host with activation key and
force such process.

pietro committed:
Fix versionadded on math filters section (#20674)

Extra character was breaking the rendering of the file.

privateip committed:
one more bug fix for ansible-connection (#20707)

The connection instance will now return a meaningful error if the
connection property returns False

privateip committed:
fixes issue in network_cli where _connected wasn't set (#20708)

Adds the code to set and unset the _connected propery in network_cli
on calls to open() and close()

bcoca committed:
fix improper setting become user for tmpdir

fixes #20706

bcoca committed:
only ignore become for making tmpdir

(cherry picked from commit 65ee9d2e4629cbb31178b54831612846ca7b3b69)

dagwieers committed:
Make test/runner less verbose when spawning instances

As discussed on IRC

supertom committed:
[GCE] Google Cloud Pubsub Module (#19091)

* Google Cloud Pubsub Module

The Google Cloud Pubsub module allows the Ansible user to:
* Create/Delete Topics
* Create/Delete Subscriptions
* Change subscription from pull to push (and configure endpoint)
* Publish messages to a topic
* Pull messages from a Subscription

An accessory module, gcpubsub_facts, has been added to list topics and subscriptions.

* Added docs for state field to DOCUMENTATION and RETURN blocks.

wyscan committed:
Fixed issue in gce_net with firewall rules that have protocols that d… (#20528)

* Fixed issue in gce_net with firewall rules that have protocols that do not have ports.

* Fixed elif in gce_net to include check of the ports variable.

mattclay committed:
Run fewer tests for winrm config script changes.

JimFicarra committed:
Add feature - win_package - allow non-zero return codes - Issue #20408 (#20410)

Commit:
additional exception handling

* catches TypeError when trying to load json data

JimFicarra committed:
win_feature fix issue #20433 - source param not found on win2k8 (#20435)

privateip committed:
updates eos shared modules (#20738)

* eos module now uses network_cli connection plugin
* adds unit tests for eos module
* eapi support now provided by eapi module
* updates doc fragment for eapi common properties

hoffmann committed:
fix typo

jhawkesworth committed:
Merge pull request #20749 from hoffmann/patch-1

fix typo

privateip committed:
refactor eos_eapi module (#20740)

* eos_eapi module now requires network_cli plugin
* adds unit test cases for eos_eapi module

privateip committed:
refactors eos_command module (#20739)

* eos_command now uses network_cli connection plugin
* adds unit test cases for eos_command module

dagwieers committed:
win_shortcut: Clean up (#20514)

Small improvements:
- Remove obsolete comments
- Use $check_mode instead of my own $_ansible_check_mode
- Renamed "default" window style to "normal" (as is used by Windows
  documentation)
- A fix for check-mode support (when state=absent)

privateip committed:
refactors eos_template to use network_cli (#20742)

Shrews committed:
Merge pull request #19320 from manics/ansible-modules-core/pull/5176

Allow creation of Openstack volumes from an existing volume

dagwieers committed:
win_ping: Make full-functional

Small cleanup and backport `ping` functionality.

dagwieers committed:
win_copy: Fix for idempotency (#20281)

* include source file in error message

* win_copy: Fix for idempotency

This patch fixes an idempotency issue with win_copy. Without this patch
files would always be considered changed (unless the copy operation failed).

It also fixes the resulting output cfr. what was deocumented.

trondhindenes committed:
Merge pull request #20599 from dagwieers/win_regedit-cleanup

win_regedit: Cleanup, check/diff mode support, HKCC fix

mattclay committed:
Temporarily remove FreeBSD 11.0 from CI.

privateip committed:
new module eos_user (#20759)

* adds new module for managing local user accounts on eos devices
* adds unit test cases for module

This module only supports network_cli

privateip committed:
refactor eos_facts to use network_cli (#20760)

refactors the eos_facts module to use the network_cli plugin

abadger committed:
Combine jimi-c and bcoca's ideas and work on hooking module-utils into PluginLoader.

This version just gets the relevant paths from PluginLoader and then
uses the existing imp.find_plugin() calls in the AnsiballZ code to load
the proper module_utils.

Modify PluginLoader to optionally omit subdirectories (module_utils
needs to operate on top level dirs, not on subdirs because it has
a hierarchical namespace whereas all other plugins use a flat
namespace).

Rename snippet* variables to module_utils*

Add a small number of unittests for recursive_finder

Add a larger number of integration tests to demonstrate that
module_utils is working.

Whitelist module-style shebang in test target library dirs

Prefix module_data variable with b_ to be clear that it holds bytes data

mattclay committed:
Temporarily remove Windows Server 2016 from CI.

mattclay committed:
Add `pep8` to `ansible-test`. (#20745)

The rule sets and legacy file list are a first draft.
It is likely that they will need to be revised.

mattclay committed:
Fix infrequent PEP 8 issues.

robwagner33 committed:
Created gce_snapshot module (#20057)

* Added support for a gce_snapshot module for managing snapshots in the Google Cloud Platform

* Fixed using sys.exit in favor of fail_json. Followed the pattern in gce_pd

* Fixed using sys.exit in favor of fail_json. Followed the pattern in gce_pd

* Fixed an issue when dealing with multiple disks. Added a bit more documentation explaining the disks argument

* Corrected a typo in the multi-disk snapshot example

* Updated module to return list of changed snapshots instead of just a message. Added a RETURN block

tedder committed:
add variable precedence warning to set_fact documentation. (#20774)

* add variable precedence warning to documentation.

Not sure how to link to other documentation.

* fix wording, thx bcoca

privateip committed:
refactors eos_config module to use network_cli (#20741)

* update eos_config to use eapi exclusively and remove cli transport
* add unit test cases for eos_config
* updates action plugin to handle both eapi and network_cli connections

dharmabumstead committed:
Refurbish developing modules content - stage 1 (#20673)

* Revamping module development docs - work in progress

* Refurb of developing modules content continues.

* Developing modules refurb work continues

* Continued refurb of developing modules content. Work-in-progress.

* Ibid

* Dev guide content refurb continues - WIP

* Ibid.

* Removed reference to old extras module repo

* Tweaks

* Removed some non-intro material; added link to github module PRs.

* Incorporated review feedback from @gundalow and @tkuratomi; fixed some links; renamed '*contributing' to '*checklist'

dagwieers committed:
Ensure that the script is run with elevated privileges (#20669)

* Ensure that the script is run with elevated privileges

This fixes #20654

* Implement our own check for elevated privileges

nnutter committed:
fix exception when an image has no name

This prevents an exception from occurring when an image has no name.  While images normally have names it is possible, at least on older versions of Docker, for an image to "lose" its name during a failed `docker rm`.

dagwieers committed:
powershell.ps1: Use Get-AnsibleParam and more (#20518)

- Harmonize the name $params like most modules
- Use `return` consistently for return values (easier to read)
- Implement Get-AnsibleParam internally using -type "bool"
- Use a dictionary for $result

jborean93 committed:
Feature/win stat extra info (#19148)

* Added more return results to win_stat

* Changed Win2012 methods to support older versions in setup

* staging of the tests to work with older servers

dagwieers committed:
Removing an non-existing directory complains (#19014)

The following playbook:

```yaml
- hosts: localhost
  tasks:
  - file:
      path: /tmp/non-existing-foo-bar
      state: absent
      recurse: yes
```

causes this error:

```
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "recurse option requires state to be 'directory'", "path": "/tmp/non-existing-foo-bar", "state": "absent"}
```

The included fix ensures that when recurse is added, we no longer assume
it is a file, but accept that it is a directory.

dagwieers committed:
Add -type "path" to path and creates parameter (#20402)

I intended to update the documentation, but it appears the website docs
are behind the actual documentation (as the website doesn't show the
default values for booleans).

So I ended up adding the missing -type "path" and use the default $null
value for undefined parameters.

mattclay committed:
PEP 8 W291 whitespace cleanup.

dagwieers committed:
win_uri: module cleanup and added check-mode support (#20511)

The following changes have been made:
- Added check-mode support
- Added parameter type support
- Replace PSObject by a normal dictionary
- Improve idempotency (only when $dest is provided it is not idempotent (yet)

dagwieers committed:
win_file: Add check-mode support (#20671)

Also a small cleanup

mattclay committed:
PEP 8 whitespace cleanup. (#20783)

* PEP 8 E271 whitespace cleanup.
* PEP 8 W293 whitespace cleanup.
* Fix whitespace issue from recent PR.

isopy committed:
default voice of Alex seems safer (#20718)

mattclay committed:
PEP 8 cleanup. (#20789)

* PEP 8 E703 cleanup.
* PEP 8 E701 cleanup.
* PEP 8 E711 cleanup.
* PEP 8 W191 and E101 cleanup.

mattclay committed:
PEP 8 cleanup. (#20790)

* PEP 8 E115 cleanup.
* PEP 8 E131 cleanup.

privateip committed:
reorganizes eos shared module (#20795)

* adds send_config function to handle pushing configuration commands
* adds support for handling multiline banner commands
* adds env var check to enable or disable config sessions

privateip committed:
adds module eos_banner (#20796)

* new module allows for configuration of eos multiline banners
* adds unit test cases for module

privateip committed:
adds new module eos_system (#20797)

* provides declarative resource for configuring eos system parameters
* supports network_cli connection plugin only
* adds unit tests for module using mock interfaces

mattclay committed:
Remove duplicate import.

hloeffler committed:
add META Output for -vv, better than nothing

mattclay committed:
PEP 8 indent cleanup. (#20800)

* PEP 8 E121 cleanup.

* PEP 8 E126 cleanup.

* PEP 8 E122 cleanup.

tedelhourani committed:
Add bigmon policy module (#20584)

* Cleanup fork

* Add bigmon_policy module

* Remove docker files from commit, fix formatting, add CHANGELOG.md entry

* Put change in CHANGELOG.md in alph order and add the alias to access_token doc

* Fix access_token type, drop the alias, and fix doc

* Capitalize letter, delete redundant notes, make validates_certs True by default

* Add validate_certs to example

* Try fixing trailing space

* Try adding a trailing space

* Try removing trailing space at end of file

* Add newline at end of file and fix trailing space

gundalow committed:
No more split repos (#20804)

* All bugs go to into `ansible/ansible`
* Issue template is automatically added when raising a new issue, so no manual steps needed

jsumners committed:
Add sv module to manage runit services #20181 (#20471)

* Add sv module to manage runit services

* Add mission `restarted` method

bcoca committed:
renamed sv to runit, avoid svc confusion

bcoca committed:
added runit and daemontools to service mgr detect

Commit:
updates python2.4-skip

* adds module_utils/netconf.py
* adds module_utils/junos.py

bcoca committed:
improved openrc detection

dagwieers committed:
Show remote host in error message. (#20823)

* Show remote host in error message.

* Typos :-/

rcarrillocruz committed:
Refactor iosxr_command to make use of network_connection plugin (#20772)

abadger committed:
Fix import of urlparse on python3

Should fix the error reported here:
https://github.com/ansible/ansible/issues/17495#issuecomment-267921719

abadger committed:
Fix service's exec_command() for python3 (#20826)

* Fix service's exec_command() for python3

exec_command() was mixing text and bytes in several places on python3.
Made changes so that we explicitly convert between the two.

Fixes #20818

* Also handle basestring and shlex.split in a python2/python3 compatible fashion

andmos committed:
Example of usage in znode documentation #20816

Commit:
Ansible Tower organization module (#20355)

* add Ansible Tower organization module

* skip Python 2.4 check for ansible_tower module

* make spec and doc match, extract tower auth helper method

* added auth params at module level

* support check mode

* extract check mode check to ansible_tower utils, add utils to 2.4 skip

* update interpreter shebang

* remove colon from docs

* no log for password, verify_ssl default to true

bcoca committed:
updated friendlier description

bcoca committed:
added tower_org

jasperla committed:
Add support for additional properties when creating new VMs. (#20792)

* Add support for additional properties when creating new VMs.

All vmadm(1M) options except those specific to SDC(-docker)
are now supported.

* Mark password options as no_log

Change-Id: I9dfd80e7ba5ecfc8f45ecfbe9ec098e66647e9a4

logan2211 committed:
os_network: Allow freeform provider_network_type string

Prescribing types is not necessary as the underlying shade library
does not do so, and the Neutron API will inform us if a disallowed
or non existent type is used.

Fixes #20830

abadger committed:
Fix for swupd documentation

bcoca committed:
removed module merged by mistake with facts change

should be it's own PR

bcoca committed:
fix cowsay for py3

Shrews committed:
Merge pull request #20832 from Logan2211/os-network-provider-types

os_network: Allow freeform provider_network_type string

ryansb committed:
Convert error to the native text type before writing to stderr

Fixes #20588

abadger committed:
Unittests for some of module_common.py (#20812)

* Unittests for some of module_common.py
* Port test_run_command to use pytest-mock

The use of addCleanup(patch.stopall) from the unittest idiom was
conflicting with the pytest-mock idiom of closing all patches
automatically.  Switching to pytest-mock ensures that the patches are
closed and removing the stopall stops the conflict.

mattclay committed:
PEP 8 E111 & E114 cleanup. (#20838)

mattclay committed:
PEP 8 E111 cleanup for nxos modules.

mattclay committed:
PEP 8 E111 cleanup. (#20841)

mattclay committed:
PEP 8 E111 cleanup.

mattclay committed:
PEP 8 E111 cleanup.

mattclay committed:
PEP 8 E111 cleanup.

mattclay committed:
PEP 8 E111 cleanup.

willthames committed:
Improve boto profile handling for EC2 inventory (#20280)

Ensure command line profile argument and AWS_PROFILE environment variable
overrides config file
Remove unnecessary `lambda` function
Fix cache file path construction to be more pythonic (and windows-ready)

heni committed:
unicode is absent in case of python3 - fix it (#20861)

Tahvok committed:
Some better explanation to release cycle

thisdougb committed:
added ldap auth capability to hashi_vault plugin (#20244)

* added ldap authentication capability

sebastianneubauer committed:
s3 module: fix urlparse invocation and netloc mixup on Python 3 (#20836)

mattclay committed:
PEP 8 cleanup.

vaygr committed:
Add Source Mage GNU/Linux facts (#19671)

* Add Source Mage GNU/Linux facts

* Add unit test for OS-detection

bcoca committed:
reworded conditional skipping to not imply error

fixes #19068

gundalow committed:
iosxr tests: Initial work for 2.3 (#20883)

Ensure a loopback device exists for testing
Remove provider:
Default is to run task on Network, device, use delegate_to: localhost for local tasks.

bcoca committed:
added ec2_vpc_nat_gateway_facts to chnglog

OtherSystems committed:
Update intro_installation.rst (#20854)

Added command to install the repository key to prevent errors

alikins committed:
Rhn register misc fixes (#20724)

* Remove monkeypatching of rhn config in rhn_register

rhn_register.Rhn() was doing some complicated monkeypatching of
the up2date_client.config.Config() class to add a default config
value. Since that was only used in one place, remove the monkeypatch
and handle the single default case.

That case was 'server_url' option, so replace it with a Rhn.server_url
property. Also handle the error case when no server url is provided.

* refactoring
* flatten some indention levels
* add 'enable_eus' to module doc
* set enable_eus var and use it directly
* style/pep8/etc cleanups
* some import cleanups

types was unused, os/re were from module_utils * import

* remove * import from module_utils.redhat
* remove * imports from module_utils.base
* remove unused Rhn._subscribe method
* cleanup pep8 style stuff
* remove a unused default mutable arg from Rhn.subscribe

Rhn.subscribe() only gets called in one place, and thats
with channels from module_params which default to [] and
are typed as a list.

bcoca committed:
Warn reserved (#20567)

* removed unused vars

* added warning when using reserved vars

fixes #4816

* cleanup

dharmabumstead committed:
Edited for grammar, punctuation and spelling

mattclay committed:
PEP 8 fix.

scboucher committed:
Feature/lookup keyring (#18379)

* Add keyring lookup

* Better error control

* Update keyring.py

* Update keyring.py

typo

bcoca committed:
added example keyring looukp

bcoca committed:
added kyring to chnglog

bcoca committed:
clean up empty warnings/deprecations from modules

mattclay committed:
PEP 8 fixes.

dagwieers committed:
Fix broken Shippable

ganeshrn committed:
module_utils/basic.py: Support logical or condition in required_if (#20220)

* Support logical or condition in required_if

Add logical 'or' condition support in 'required_if'
for requirements.
* If requirements is a list all parameters within it should
  be present.
* If requirements is a set atleast one parameter should
  be present

* Fix review comment

allanice001 committed:
Update test-module (#20737)

* Update test-module

Ensuring invoke is assigned

Traceback (most recent call last):
  File "ansible/hacking/test-module", line 267, in <module>
    main()
  File "ansible/hacking/test-module", line 263, in main
    runtest(modfile, argspath, modname, module_style, interpreters)
  File "ansible/hacking/test-module", line 207, in runtest
    invoke = "%s%s" % (invoke, modfile)
UnboundLocalError: local variable 'invoke' referenced before assignment

* Update test-module

Made the change to only require a single if, making the function more 'DRY'.

ivanbojer committed:
new panos module to import software or configuration file into firewall (#19901)

* # This is a combination of 6 commits.
# The first commit's message is:
new module to import software or configuration file onto firewall

# This is the 2nd commit message:

changes based on the review comments; remove unecessary if statements; change returned value docstring

# This is the 3rd commit message:

empty checkin to trigger ANSIBLEbot

# This is the 4th commit message:

added additional exception handling

# This is the 5th commit message:

- added new module info to the changelog as requested

# This is the 6th commit message:

removed blank space as tox checks were failing

* new module to import software or configuration file onto firewall

privateip committed:
removes assert and uses to_list common function (#20866)

non impacting change to refactor assert statement out of ios shared
module

privateip committed:
moves the module utils network cli transport into a separate module (#20921)

This move will allow for the cli transport to go through a deprecation
process as we transition to network_cli plugin

n0ts committed:
ec2.py support region auto (#15427)

naslanidis committed:
[cloud] New module ec2_vpc_igw_facts (#19787)

* added ec2_vpc_igw_facts module

* added cr at end of file

* corrected import json in wrong location

* corrected version added

* added snake_case conversion

* updated documentation and fixed for python 3'

* ec2_vpc_igw_facts: simplify logic

Make module arguments more 'Ansiblish'
Remove unnecessary intermediate variables in results generation
Use `ansible_dict_to_boto3_filter_list` rather than duplicating logic
Use `check_mode` rather than pass a `dryrun` argument
Update for flake8 improvements

* updated documentation

ryansb committed:
[docs] Add ec2_vpc_igw_facts to changelog

supertom committed:
[GCE] External IP Address Module. (#20779)

* [GCE] External IP Address Module.

This module allows users to create and delete External IP Addresses. Both Regional and Global Addresses are supported.

* Removed whitespace causing pep8 issue

ryansb committed:
Add gce_eip module to changelog

jctanner committed:
Use a -short- custom hash for controlpersist path by default  (#20843)

* A method to validate and alter the ssh control path automatically.
* First tries %C to use the shortened hash
* On further failure, it removes section by section from the original path
* Fix hostname
* Implement bcoca's suggested changes
* Remove unused option
* Remove unused class var
* Use to_string to avoid unicode error
* Switch from to_text to to_bytes
* Update the example config for the new controlpath feature

jctanner committed:
Update changelog for controlpersist path feature (#20928)

ganeshrn committed:
Fix update check issue (#20915)

Add one of condition in required_if
to ensure configuration statements are
present if update value in provided in
playbook.

mattclay committed:
PEP 8 fix.

abadger committed:
Move ssh and local connection plugins from using raw select to selectors

At the moment, this change will use EPoll on Linux, KQueue on *BSDs,
etc, so it should alleviate problems with too many open file
descriptors.

* Bundle a copy of selectors2 so that we have the selectors API everywhere.
* Add licensing information to selectors2 file so it's clear what the
  licensing terms and conditions are.
* Exclude the bundled copy of selectors2 from our boilerplate code-smell test
* Rewrite ssh_run tests to attempt to work around problem with mocking
  select on shippable

Fixes #14143

resmo committed:
module_utils: implement deprecation warning for params (#20884)

* module_utils: implement deprecation warning for params

* rename deprecated_version to removed_in_version

* fix pep8 E121

bcoca committed:
allow device to be list for multidev fs (#20655)

* allow device to be list for multidev fs

fixes #20551

* reverted command to string

bcoca committed:
last ditch debian release data (#20571)

fixes #10445

mattclay committed:
Add `--debug` option to `ansible-test`.

maximede committed:
Fix for #20473 - If there is no instance in the ASG, we initialize an empty array for `properties['instances']` (#20475)

willthames committed:
Fix EIP release in ec2_vpc_nat_gateway (#20167)

* Check if EIP exists before deleting it

After deleting the NAT gateway, the EIP sometimes seems to
cease to exist afterwards. Check if it exists before deleting it.

Otherwise you get:
```
Failed to release EIP eipalloc-abdc1234: An error occurred (InvalidAllocationID.NotFound) \
when calling the ReleaseAddress operation: The allocation ID 'eipalloc-abcd1234' does not \
exist", "success": false}
```

* Fix flake8 errors with ec2_vpc_nat_gateway

willthames committed:
Disassociate EC2 VPC subnets from route tables before deletion (#20114)

* Disassociate subnets from route tables before deletion

If a route table still has subnets associated with it, it will fail
to delete:

```
"msg": "The routeTable 'rtb-abcd1234' has dependencies and cannot be deleted."
```

Avoid this by disassociating subnets before route table deletion

* Fix ec2_vpc_route_table flake8 complaints

bcoca committed:
added back return false for when not debian

mattclay committed:
Update PEP 8 legacy files list.

abadger committed:
Do not substitute ssh_exeuctable until we need to

We need to use ssh_executable instead of hardcoding ssh in the command
we run but we need to use "ssh" when we lookup the value of the
{command}_extra_args variable.  Do this by leaving binary as "ssh" and
only expanding when we place it into b_command.

Fixes #20862

privateip committed:
fixes exec_command in shell (#20948)

transforms command to dict for processing by instance of Shell

privateip committed:
adds eos_local shared module to handle local connections (#20960)

* adds support for eapi
* adds support for cli
* adds doc fragement

privateip committed:
adds support for eos_local connections to eos modules (#20961)

* updates modules to use eos_local

Commit:
clean up unused doc fragments

privateip committed:
adds the cli transport back to the ios modules (#20949)

the cli transport was initially removed to aid in the transition to
network_cli.  This adds the function back to the provider can be
specified.

Commit:
Revert "clean up unused doc fragments"

Those "unused" doc fragments are still referenced

lib/ansible/modules/network/eos/eos_config.py:extends_documentation_fragment: eapi
lib/ansible/modules/network/eos/eos_facts.py:extends_documentation_fragment: eos

This reverts commit 246cd041d8e68045c64db7a2d8c1d14b9e7e4233.

dagwieers committed:
win_shortcut: Add missing $check_mode definition + bugfix + tests (#20911)

* win_shortcut: Add missing $check_mode definition

For some reason this entry was missing, possible a merge-conflict gone
wrong :-(

* Added integration tests and bugfix

Add missing changes.

gundalow committed:
Roadmap is draft (#20970)

The keyword is "draft",  not "final".

simplesteph committed:
fixed type mapper and added tests (#20963)

s-hertel committed:
[cloud] rds module: handle parameters that are False, but also not the default value. Fixes #20370 (#20646)

Boolean options that default as `None` but are set to `False` by the user were ignored on update. This change checks to distinguish None & False so that options like multi_az can be turned off during an update. 

* Modifying how optional parameters are handled in rds.py. Fixes #20370

Allowing options to be set to false/no. Previously ignored unless set to true/yes.

Added a conditional for invalid parameters since the default is false instead of null for some options (e.g. force_failover, apply_immediately, upgrade).

* Making requested revision.

privateip committed:
fix up eos doc frags (#20981)

* fix modules with invalid doc frags
* remove unneeded doc frags

Qalthos committed:
vyos_config 2.3 (#20577)

* Rename vyos2 over vyos

* Update vyos_config to LocalAnsibleModule

Change result key 'updates' -> 'commands'

vyos_config is supported by core

* vyos_config tests

* Enable bracket config use

* Sanitize config before use

bcoca committed:
allow debug to be 'changed' (#20957)

* allow debug to be 'changed'

fixes #14430

* removed test that checked against this

machacekondra committed:
ovirt: Add support to diff (#20698)

* cloud: ovirt: Add diff support to

* cloud: ovirt: Add ability to print nested list of objects

* cloud: ovirt: Fix waiting while reinstalling host

* cloud: ovirt: fix ovirt_vms 404 error

* cloud: ovirt: add check_mode support to ovirt_auth module

machacekondra committed:
Ovirt vms improve (#20882)

* cloud: ovirt: add function to get id by name

* cloud: ovirt: add instance type parameter

* cloud: ovirt: use param method instead of module.params

* cloud: ovirt: use 'and' at begging of next line

* cloud: ovirt: add description parameter to vms module

* cloud: ovirt: add comment parameter to vms module

* cloud: ovirt: add timezone parameter to vms module

* cloud: ovirt: add serial_policy parameter to vms module

Commit:
univention udm_user: add parameter to control password update (#20699)

* univention udm_user: added support to set the password always

* module udm_user: added version 2.3 to update_password

privateip committed:
fixes handle_prompt in shell module (#20982)

* transforms command objects to dict
* transforms string commands to dict

Qalthos committed:
vyos_facts 2.3 (#20939)

* Update vyos_facts documentation

* Replace runner for run_commands

* vyos_facts tests

* Update documentation

* vyos_facts is supported by core

sivel committed:
Update validate-modules (#20932)

* Update validate-modules

* Validates ANSIBLE_METADATA
* Ensures imports happen after documentation vars
* Some pep8 cleanup

* Clean up some left over unneeded code

* Update modules for new module guidelines and validate-modules checks

* Update imports for ec2_vpc_route_table and ec2_vpc_nat_gateway

ivanbojer committed:
added already approved PANOS modules (#20952)

machacekondra committed:
cloud: ovirt: add host engine options to ovirt_hosts (#20967)

machacekondra committed:
[cloud][ovirt] new module ovirt_snapshots & ovirt_snapshots_facts (#20936)

* cloud: ovirt: add snapshots module

* Move imports in ovirt_snapshots module to match style & pass CI

* Move ovirt_snapshot_facts imports to comply w/ CI

ryansb committed:
Add ovirt_snapshots to changelog

abadger committed:
Split on newlines when checking for prompt matches (#20945)

* Check for the prompt as a substring of the output

sudo sometimes spits out warnings to stdout before getting to the
password prompt.  Account for that when trying to match a password
prompt.

Fixes #20858

rektide committed:
Update loop docs to also describe register:'s value when inside loop. (#20991)

privateip committed:
bug fixes for shell exec command (#20988)

* fixes up eos_local to call appropriate method
* fixes up ios_cli to call appropriate method
* moves json conversion to eos

mattclay committed:
Fix exception in module validator.

abadger committed:
Module utils default path (#20913)

* Make the module_utils path configurable
* Add a config value to define the path site module_utils files
* Handle module_utils that do not have source as an error
* Make an integration test for module_utils envvar working
* Add documentation for the ANSIBLE_MODULE_UTILS config option/envvar
* Add it to the sample ansible.cfg
* Add it to intro_configuration.
* Also modify intro_configuration to place envvars on equal footing with
  the config options (will need to document the envvar names in the
  future)
* Also add the ANSIBLE_LIBRARY use case from
  https://github.com/ansible/ansible/issues/15432 so we can close out
  that bug.

resmo committed:
ios_cli: fix load_config exec_command (#20994)

fixes AttributeError: 'AnsibleModule' object has no attribute 'exec_command'

privateip committed:
fixes ios and eos command modules (#20989)

* returns support for prompt/response over cli
* now sends native dict instead of str command
* fixes issue with run_commands() in ios to jsonify request
* updates unit test cases

privateip committed:
remove unneeded eos_cli shared module (#20999)

Commit:
Update tox.ini to match current standards

tox.ini is used for flake8 configuration in addition to tox
configuration. Update the flake8 configuration to match the
latest standards as documented in ansible/proposals#50

jhawkesworth committed:
Fix 'touch' mode so it creates zero size files. (#20876)

abadger committed:
Make configparser imports python3 ready

the ConfigParser module was renamed to configparser in Python3.  Use
six.moves to import it so that the modules will function on Python3.

dav1x committed:
closes #21011 (#21015)

gundalow committed:
make webdocs speedup is done

mattclay committed:
Fix win_stat test for changes from PR #20876

The changes to win_file in that PR resulted in changes to file
sizes and file hashes expected by the win_stat tests.

alikins committed:
Add 'Improve testing of setup.py/facts.py' to 2.4 roadmap

bcoca committed:
added docs for vault and made trigger shorter: !vault (#20985)

* added docs for vault and made trigger shorter: !vault

* added single var valuting

* Update playbooks_vault.rst

Edit pass for spelling and grammar. Ship it!

* Update playbooks_vault.rst

Typo fixes.

ryansb committed:
[cloud][aws] Make `s3_object_version` truly optional in lambda module

mattclay committed:
Fix @contextmanager leak on exception. (#21031)

* Fix @contextmanager leak on exception.
* Fix test leaks of global module args cache.

abadger committed:
Make sure that we're comparing text all the way through.

On Darwin, sys.platform returns byte strings on both python2 and
python3.  Turn it into a text string everywhere in order to remedy that.

Fixes #19845

dagwieers committed:
Improve inline docs (#21029)

speshak committed:
Correct typo in net_command docs (#20976)

JimFicarra committed:
win_msi:Fixing Issue #20315 - shows changed even when msi not found (#20358)

* win_msi #20315 - shows changed even when msi not found

* make brace placement consistent

ganeshrn committed:
Fixes #20412 junos_config delete command fix (#20450)

If same config hierarchy is set and deleted in one playbook
for delete statement add support to check if the config is present
on device or in the playbook. If it is present add delete statement in
updated config list.

privateip committed:
adds netconf shared module (#20743)

New netconf shared lib added to module_utils for sending netconf
messages to remote devices over netconf connection plugin

privateip committed:
fixes minor issues with shell (#21039)

* adds sendonly to command dict
* fixes setting timeout param

dagwieers committed:
Updated CHANGELOG.md with previous bits (#20973)

And also sort the modules properly.

dagwieers committed:
Add missing support for -CertValidityDays (#21009)

* Add missing support for -CertValidityDays

For some reason the -CertValidityDays option was not being used in the certificates we created.

This fixes #10439

* Possible fix

* We cannot use New-SelfSignedCertificate on 2012R2 and earlier

As suggested by @jhawkesworth

dminca committed:
update vague description of only_upgrade (#21050)

* the `only_update` param description was kinda vague:
  * either install if not installed
  * or upgrade if installed
  * or both!?
    * as a result, remove the "install" statement from this sentence

Resolves:
Related:
Signed-off-by: Daniel Andrei Minca <mandrei17@gmail.com>

t0mk committed:
Dynamic inventory for Packet host (https://packet.net) (#20635)

* Add dynamic inventory script and config for Packet.net

* The script and config have been shamelessly cargo
  culted from the `ec2.py` and `ec2.ini` dynamic inventory
  script.
* This is an initial version and could very well be
  enhanced and made better.

Examples:

`PACKET_NET_API_KEY=<MY_AUTH_TOKEN> --list` to get inventory for
all hosts in Packet.net in all projects (defaults to `--list`
if no argument provided).

`PACKET_NET_API_KEY=<MY_AUTH_TOKEN> --host HOST` to get variables
for a single host.

* improvements in Packet host dynamic inventory

gundalow committed:
module_utils/vmware.py is now clean

gundalow committed:
Update legacy-files.txt

sivel committed:
pep8 fixes for recent packet_net inventory (#21066)

abadger committed:
Fix typo when choosing KqueueSelector

nerzhul committed:
New modules: vmware_guest_facts & vmware_guest_snapshot (#20971)

* Move gather_facts action outside of vmware_guest module
* Move facts gathering method to vmware.py library
* vmware_guest use some vmware.py library methods
* vmware.py lib: fix some PEP 8 issues
* Move snapshot actions to vmware_guest_snapshot
* Move common snapshot facts to vmware.py library

bcoca committed:
minor fixes to template docs

imranh2 committed:
Add Check Mode capability to kernel_blacklist module (#21007)

* Add Check Mode capability to kernel_blacklist module
* Add suggested changes from @tmshn
* Pass a bool into `Blacklist` that'll just be `module.check_mode`
* Move detection and creation of a file to a separate function within `Blacklist`
* If there's no file on the system and we are running under `Check Mode`, set `self.filename` to `os.devnull` AND mark a change, as the module would have created an empty file.
* Whenever a `self.filename` is being opened in a mode where changes can be made, replace with an if statement that checks `self.checkmode` and if true then open up `os.devnull` instead

evgeni committed:
use require_if to check for username or AK instead of custom code (#20937)

the logical or parameter to require_if was introduced in #20220

dav1x committed:
fixes 20904 (#21068)

bcoca committed:
error out if systemctl isn't present

heni committed:
fix urlparse usage in cloud.amazon module to be compatible with python3 (#20917)

* fix urlparse usage in cloud.amazon module to be compatible with python3

* make imports style compatible with six documentation

abadger committed:
Cleanup imports

rcarrillocruz committed:
Refactor iosxr_facts to use network_cli plugin (#20857)

vincentbernat committed:
cloudstack: ensure we select a template compatible with disksize (#21058)

When the user specify a root disk size, the template image must be
smaller than the specified root disk size.

SaaldjorMike committed:
Fix indentation to match other examples (#21080)

fengor committed:
improve error message on conflicting actions. Fixes #21057 (#21059)

* this fixes #21057

outputs the conflicting action statements.

* reverting to old  % syntax for python 2.6 compatibility

mattclay committed:
Use us-east-2 for Windows CI on Shippable.

mattclay committed:
Update Testing and CI section of 2.3 roadmap.

privateip committed:
starts cli if root user login is used for junos terminal (#20761)

fixes #20131

jtyr committed:
Making the decoration more clear (#20931)

craigfrancis committed:
Reminder that lookup(‘file’) can be used (#20649)

* Reminder that lookup(‘file’) can be used

Sometimes the block of text does not easily fit into a playbook, so this acts as a reminder (or a prompt for anyone who hasn't used this before) that the content could be stored in an separate file.

I've also included `backup: yes` as I think this is a good example of where a backup might be needed.

* Correct the not so "smart quotes"

* Update blockinfile.py

Replace – with an an ASCII -

gundalow committed:
developing_module status

Abdul-Anshad-A committed:
Add guest shutdown/reboot (#20564)

ryansb committed:
Update roadmap for cloud work

sivel committed:
Legacy pep8 clean fixes for contrib and hacking (#21081)

mckerrj committed:
Update ROADMAP_2_3.rst

mckerrj committed:
Update ROADMAP_2_3.rst

mckerrj committed:
Update ROADMAP_2_3.rst

gundalow committed:
formatting

alikins committed:
update ansible-ssh in roadmap

gundalow committed:
formatting

alikins committed:
More unit tests for plugins/action/ (#20887)

Qalthos committed:
Vyos system module (#21060)

* vyos_system tests

* Merge vyos_system fixture with vyos_config fixture

alikins committed:
Redhat subscription (#20729)

* Make unused redhat_subscriptions do something

rhsm_baseurl/server_insecure were module params that were
never used previously. Hook them up for register options.

* pep8/style cleanups
* Set a module scope SUBMAN_CMD before we ref it as a global
* move ansible imports to top, remove '*' imports
* remove redhat_subscriptions from pep8 legacy

nitzmahone committed:
update azure roadmap items

alikins committed:
Fix action test (#21102)

* update _parse_returned_data for no _ansible_parsed

_parse_returned_data no longer adds _ansible_parsed

* pep8 cleanups

gundalow committed:
Update ROADMAP_2_3.rst

bverschueren committed:
fix missing ident in provider dict assignment (#21089)

Thanks @bverschueren!

rcarrillocruz committed:
Set block indentation to 1 on iosxr templates (#21105)

Network devices in the lab have by default an indentation of 1 for
sub-sections, and so does the netcfg.NetworkConfig constructor
indent param.
This will fix reported issue 21055, and make the integration tests
to pass.

Fixes #21055

s-hertel committed:
Two added AWS modules for ElastiCache

resmo committed:
cs_instance: fix ssh keys not found in projects (#21104)

Follow up of 5475f3ee644c32d03f48fb091ef019b7b5fd2706

mattclay committed:
Fix vyos unit tests.

abadger committed:
Add 2.3.0 entries for module_utils via pluginloader, selectors, and py3 (#21019)

* Add 2.3.0 entries for module_utils via pluginloader, selectors, and py3

* Add roadmap-2.3 updates and an item for roadmap-2.4

DavidVentura commented:
I totally forgot about this. What do we do now? Do you want me to rebase this ?

dagwieers committed:
Add missing iteritems definition (#21106)

Also moved most imports out of try-except block.

gundalow commented:
Please rebase

gundalow commented on lib/ansible/modules/cloud/misc/proxmox.py:
```diff
@@ -175,6 +175,14 @@
     default: present
+  pubkey:
+    description:
+      - Public key to add to /root/.ssh/authorized_keys.
```
If I'm reading the code correct, does this only work on proxmox 4.2+? If so may be worth stating that here.

gundalow commented on lib/ansible/modules/cloud/misc/proxmox.py:
```diff
@@ -175,6 +175,14 @@
+    version_added: "2.3"
+    default: null
+    required: false
+    type: string
```
No need to document `type:`, please remove.

DavidVentura committed:
Merge branch 'proxmox-pubkey-support' of github.com:DavidVentura/ansible into proxmox-pubkey-support

DavidVentura committed:
fix whitespace, add comment about version support

DavidVentura commented:
I seem to have broken my merge beyond human ability. I'll clean this up and open a new PR with a clean branch.
